compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+void verifyPrototype(FnSymbol* fn);
+Type* getTypeFromFormal(ArgSymbol* as);
+Type* getTypeFromFormal(FnSymbol* fn, int i);
+bool typeRequiresAllocation(Type* t);
+
+private:
+
+std::string genMarshalBodyPrimitiveScalar(Type* t, bool out);
+std::string genMarshalBodyStringC(Type* t, bool out);
+std::string genComment(const char* msg, const char* pfx="");
+std::string genNote(const char* msg);
+std::string genTodo(const char* msg);
+std::string genHeaderInc(const char* header, bool system=false);
+std::string genMarshalRoutine(Type* t, bool out);
+std::string genMarshalPushRoutine(Type* t);
+std::string genMarshalPullRoutine(Type* t);
+std::string genServerDispatchSwitch(const std::vector<FnSymbol*>& fns);
+std::string genFuncToSetServerGlobals(void);
+std::string genDebugPrintCall(FnSymbol* fn, const char* pfx="");
+std::string genDebugPrintCall(const char* msg, const char* pfx="");
+std::string genFuncNumericID(FnSymbol* fn);
+std::string genServerWrapperCall(FnSymbol* fn);
+std::string genClientsideRPC(FnSymbol* fn);
+std::string genServersideRPC(FnSymbol* fn);
+std::string genMarshalCall(const char* s, const char* v, Type* t, bool out);
+std::string genMarshalPushCall(const char* s, const char* v, Type* t);
+std::string genMarshalPullCall(const char* s, const char* v, Type* t);
+std::string genTypeName(Type* t);
+std::string genSocketCall(const char* s, const char* v, const char* l,
+                          bool out);
+std::string genSocketCall(const char* s, const char* v, bool out);
+std::string genSocketPushCall(const char* s, const char* v);
+std::string genSocketPullCall(const char* s, const char* v);
+std::string genAddressOf(const char* var);
+std::string genAddressOf(std::string& var);
+std::string genSizeof(const char* var);
+std::string genSizeof(std::string& var);
+std::string genNewDecl(const char* t, const char* n);
+std::string genNewDecl(Type* t, const char* n);
+
+};
+
+//
+// This is the main entrypoint for MLI code generation, call this if the
+// necessary conditions are met in codegen().
+//
+void codegenMultiLocaleInteropWrappers(void) {
+  Vec<ModuleSymbol*> &mds = allModules;
+
+  // Insert all kinds of debug printlines into generated code for now.
+  MLIContext mli(true);
+
+  mli.emitClientPrelude();
+  mli.emitServerPrelude();
+
+  forv_Vec(ModuleSymbol, md, mds) {
+    mli.emit(md);
+  }
+
+  mli.emitMarshalRoutines();
+  mli.emitServerDispatchRoutine();
+
+  return;
+}
+
+MLIContext::MLIContext(bool injectDebugPrintlines, bool separateHeaders) {
+
+  // Yes, I know this isn't the most optimal way to initialize these!
+  this->injectDebugPrintlines = injectDebugPrintlines;
+  this->separateHeaders = separateHeaders;
+
+  openCFile(&this->fiMarshalling, gen_mli_marshalling, "c");
+  openCFile(&this->fiClientBundle, gen_mli_client_bundle, "c");
+  openCFile(&this->fiServerBundle, gen_mli_server_bundle, "c");
+  this->info = gGenInfo;
+
+  return;
+}
+
+MLIContext::~MLIContext() {
+
+  // Now, turn beautify ON!
```

Comment:
For a long time, I had beautify off and was using my own formatting. That was just a personal comment emphatically declaring that the "time for handwritten formatting" was at an end.

Reply:
Okay, but the compiler is not really the place for such comments - code here should aim more for being helpful than for amusement, since other people may have to actively maintain its contents going forward and since it is a product for wider distribution.  We occasionally leave notes in there with our own names attached but that's entirely so people know who they could contact if something goes wrong.

Reply:
Sorry, don't mean to be a killjoy

Reply:
As you say.